**Implementation of Simple ML model**

Name - Khemraj Gupta

Reg no. - 20MAI0079

In [7]:
import pandas as pd
import numpy as np
import sklearn.datasets

In [8]:
cancer_ds = sklearn.datasets.load_breast_cancer()

In [9]:
# Let's organise as X and y
X = cancer_ds.data
y = cancer_ds.target

In [10]:
print(X)
print(y)

[[1.799e+01 1.038e+01 1.228e+02 ... 2.654e-01 4.601e-01 1.189e-01]
 [2.057e+01 1.777e+01 1.329e+02 ... 1.860e-01 2.750e-01 8.902e-02]
 [1.969e+01 2.125e+01 1.300e+02 ... 2.430e-01 3.613e-01 8.758e-02]
 ...
 [1.660e+01 2.808e+01 1.083e+02 ... 1.418e-01 2.218e-01 7.820e-02]
 [2.060e+01 2.933e+01 1.401e+02 ... 2.650e-01 4.087e-01 1.240e-01]
 [7.760e+00 2.454e+01 4.792e+01 ... 0.000e+00 2.871e-01 7.039e-02]]
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 1 0 0 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 0 1 0 0
 1 0 1 0 0 1 1 1 0 0 1 0 0 0 1 1 1 0 1 1 0 0 1 1 1 0 0 1 1 1 1 0 1 1 0 1 1
 1 1 1 1 1 1 0 0 0 1 0 0 1 1 1 0 0 1 0 1 0 0 1 0 0 1 1 0 1 1 0 1 1 1 1 0 1
 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 0 1 1 0 0 1 1 0 0 1 1 1 1 0 1 1 0 0 0 1 0
 1 0 1 1 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 1 0 0 1 1
 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 0 1 1 1 1 0 1 1 1 1 1 0 1 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 1 1 1 1 1 1 0 1 0 1 1 0 1 1 0 1 0 0 1 1 1 1 1 1 1 1 

In [11]:
print(X.shape, y.shape)

(569, 30) (569,)


In [12]:
# Get the dataset inside a dataframe
data = pd.DataFrame(cancer_ds.data, columns = cancer_ds.feature_names)

In [13]:
# rename the y variable as Target 
data['class'] = cancer_ds.target

In [14]:
# Examine the first five records
data.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  class  
0          0.4601                  0.11890      0  
1          0.2750                  0.08902      0  
2          0.3613                  0.08758      0  
3          0.6638                  0.17300      0  
4          0.2364                  0.07678      0  

[5 rows x 31 columns]

In [15]:
data.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst texture  \
count     569.000000              569.000000  ...     569.000000   
mean        0.181162                0.062798  ...      25.677223   
std         0.027414                0.007060  ...       6.146258   
min         0.106000                0.049960  ...      12.020000   
25%         0.161900                0.057700  ...      21.080000   
50%         0.179200                0.061540  ...      25.410000   
75%         0.195700                0.066120  ...      29.720000   
max         0.304000                0.097440  ...      49.540000   

       worst perimeter   worst area  worst smoothness  worst compactness  \
count       569.000000   569.000000        569.000000         569.000000   
mean        107.261213   880.583128          0.132369           0.254265   
std          33.602542   569.356993          0.022832           0.157336   
min          50.410000   185.200000          0.071170           0.027290   
25%          84.110000   515.300000          0.116600           0.147200   
50%          97.660000   686.500000          0.131300           0.211900   
75%         125.400000  1084.000000          0.146000           0.339100   
max         251.200000  4254.000000          0.222600           1.058000   

       worst concavity  worst concave points  worst symmetry  \
count       569.000000            569.000000      569.000000   
mean          0.272188              0.114606        0.290076   
std           0.208624              0.065732        0.061867   
min           0.000000              0.000000        0.156500   
25%           0.114500              0.064930        0.250400   
50%           0.226700              0.099930        0.282200   
75%           0.382900              0.161400        0.317900   
max           1.252000              0.291000        0.663800   

       worst fractal dimension       class  
count               569.000000  569.000000  
mean                  0.083946    0.627417  
std                   0.018061    0.483918  
min                   0.055040    0.000000  
25%                   0.071460    0.000000  
50%                   0.080040    1.000000  
75%                   0.092080    1.000000  
max                   0.207500    1.000000  

[8 rows x 31 columns]

In [16]:
# How many malignant and benign ?
print(data['class'].value_counts()) #Binary Variable

1    357
0    212
Name: class, dtype: int64


In [17]:
# Know the target names
print(cancer_ds.target_names)

['malignant' 'benign']


In [18]:
data.groupby('class').mean()

mean radius  mean texture  mean perimeter   mean area  mean smoothness  \
class                                                                           
0        17.462830     21.604906      115.365377  978.376415         0.102898   
1        12.146524     17.914762       78.075406  462.790196         0.092478   

       mean compactness  mean concavity  mean concave points  mean symmetry  \
class                                                                         
0              0.145188        0.160775             0.087990       0.192909   
1              0.080085        0.046058             0.025717       0.174186   

       mean fractal dimension  ...  worst radius  worst texture  \
class                          ...                                
0                    0.062680  ...     21.134811      29.318208   
1                    0.062867  ...     13.379801      23.515070   

       worst perimeter   worst area  worst smoothness  worst compactness  \
class                                                                      
0           141.370330  1422.286321          0.144845           0.374824   
1            87.005938   558.899440          0.124959           0.182673   

       worst concavity  worst concave points  worst symmetry  \
class                                                          
0             0.450606              0.182237        0.323468   
1             0.166238              0.074444        0.270246   

       worst fractal dimension  
class                           
0                     0.091530  
1                     0.079442  

[2 rows x 30 columns]

In [19]:
data.groupby('mean symmetry').mean()

mean radius  mean texture  mean perimeter  mean area  \
mean symmetry                                                         
0.1060               11.20         29.37           70.67      386.0   
0.1167               11.41         14.92           73.53      402.0   
0.1203               12.05         22.72           78.75      447.8   
0.1215               12.49         16.85           79.19      481.6   
0.1220               17.85         13.23          114.60      992.1   
...                    ...           ...             ...        ...   
0.2655               24.25         20.20          166.20     1761.0   
0.2678               11.80         16.58           78.99      432.0   
0.2743               10.17         14.88           64.55      311.9   
0.2906               20.18         23.97          143.70     1245.0   
0.3040               17.14         16.40          116.00      912.7   

               mean smoothness  mean compactness  mean concavity  \
mean symmetry                                                      
0.1060                 0.07449           0.03558        0.000000   
0.1167                 0.09059           0.08155        0.061810   
0.1203                 0.06935           0.10730        0.079430   
0.1215                 0.08511           0.03834        0.004473   
0.1220                 0.07838           0.06217        0.044450   
...                        ...               ...             ...   
0.2655                 0.14470           0.28670        0.426800   
0.2678                 0.10910           0.17000        0.165900   
0.2743                 0.11340           0.08061        0.010840   
0.2906                 0.12860           0.34540        0.375400   
0.3040                 0.11860           0.22760        0.222900   

               mean concave points  mean fractal dimension  radius error  ...  \
mean symmetry                                                             ...   
0.1060                    0.000000                 0.05502        0.3141  ...   
0.1167                    0.023610                 0.06217        0.3344  ...   
0.1203                    0.029780                 0.06659        0.1194  ...   
0.1215                    0.006423                 0.05673        0.1716  ...   
0.1220                    0.041780                 0.05243        0.4834  ...   
...                            ...                     ...           ...  ...   
0.2655                    0.201200                 0.06877        1.5090  ...   
0.2678                    0.074150                 0.07371        0.3197  ...   
0.2743                    0.012900                 0.06960        0.5158  ...   
0.2906                    0.160400                 0.08142        0.9317  ...   
0.3040                    0.140100                 0.07413        1.0460  ...   

               worst texture  worst perimeter  worst area  worst smoothness  \
mean symmetry                                                                 
0.1060                 38.30            75.19       439.6           0.09267   
0.1167                 17.70            79.12       467.2           0.11210   
0.1203                 28.71            87.36       488.4           0.08799   
0.1215                 19.71            84.48       544.2           0.11040   
0.1220                 18.42           127.10      1210.0           0.09862   
...                      ...              ...         ...               ...   
0.2655                 23.99           180.90      2073.0           0.16960   
0.2678                 26.38            91.93       591.7           0.13850   
0.2743                 17.45            69.86       368.6           0.12750   
0.2906                 31.72           170.30      1623.0           0.16390   
0.3040                 21.40           152.40      1461.0           0.15450   

               worst compactness  worst concavity  worst concave points  \
mean symmetry                                    

# Train Test Split

In [20]:
from sklearn.model_selection import train_test_split

In [21]:
X = data.drop('class', axis=1)
y = data['class']

In [22]:
# Magical line of code to split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)

In [23]:
print( y.shape, y_train.shape, y_test.shape)

(569,) (512,) (57,)


In [24]:
print(y.mean(), y_train.mean(), y_test.mean())

0.6274165202108963 0.634765625 0.5614035087719298


In [25]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, stratify =  y)
print(y.mean(), y_train.mean(), y_test.mean())

0.6274165202108963 0.626953125 0.631578947368421


In [26]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2,random_state = 42) # zero or any other integer
print(y.mean(), y_train.mean(), y_test.mean())

0.6274165202108963 0.6285714285714286 0.6228070175438597


In [27]:
print(X_train.mean(), X_test.mean(), X.mean())

mean radius                 14.117635
mean texture                19.185033
mean perimeter              91.882242
mean area                  654.377582
mean smoothness              0.095744
mean compactness             0.103619
mean concavity               0.088898
mean concave points          0.048280
mean symmetry                0.181099
mean fractal dimension       0.062757
radius error                 0.402016
texture error                1.202687
perimeter error              2.858253
area error                  40.071299
smoothness error             0.006989
compactness error            0.025635
concavity error              0.032824
concave points error         0.011894
symmetry error               0.020574
fractal dimension error      0.003820
worst radius                16.235103
worst texture               25.535692
worst perimeter            107.103121
worst area                 876.987033
worst smoothness             0.131532
worst compactness            0.252742
worst concav

In [28]:
# Binarization of Input
X_binarised_train = X_train.apply(pd.cut,  bins=2, labels=[1,0])
X_binarised_test = X_test.apply(pd.cut,  bins=2, labels=[1,0])
type(X_binarised_train)

pandas.core.frame.DataFrame

In [29]:
X_binarised_train = X_binarised_train.values
X_binarised_test = X_binarised_test.values
type(X_binarised_train)

numpy.ndarray

In [30]:
b = 10

from random import randint
i = randint(0, X_binarised_train.shape[0])

if(np.sum(X_binarised_train[i,:])>=b):
    print("Model prediction is Malignant")
else :
    print("Model prediction is benign")
    
if(y_train[i]==1):
    print("Actual prediction is malignant")
else :
    print("Actual prediction is benign")

Model prediction is Malignant
Actual prediction is malignant


In [31]:
for b in range (X_binarised_train.shape[1]+1):
    accurate_rows = 0
    for x, y in zip (X_binarised_train, y_train) :
        y_pred = (np.sum(x)>=b)
        accurate_rows += (y==y_pred)
    print("b = ", b, accurate_rows, accurate_rows/X_binarised_train.shape[0])

b =  0 286 0.6285714285714286
b =  1 286 0.6285714285714286
b =  2 286 0.6285714285714286
b =  3 286 0.6285714285714286
b =  4 286 0.6285714285714286
b =  5 286 0.6285714285714286
b =  6 286 0.6285714285714286
b =  7 286 0.6285714285714286
b =  8 286 0.6285714285714286
b =  9 286 0.6285714285714286
b =  10 286 0.6285714285714286
b =  11 286 0.6285714285714286
b =  12 286 0.6285714285714286
b =  13 286 0.6285714285714286
b =  14 289 0.6351648351648351
b =  15 291 0.6395604395604395
b =  16 292 0.6417582417582418
b =  17 296 0.6505494505494506
b =  18 299 0.6571428571428571
b =  19 302 0.6637362637362637
b =  20 305 0.6703296703296703
b =  21 315 0.6923076923076923
b =  22 322 0.7076923076923077
b =  23 335 0.7362637362637363
b =  24 347 0.7626373626373626
b =  25 361 0.7934065934065934
b =  26 373 0.8197802197802198
b =  27 387 0.8505494505494505
b =  28 389 0.8549450549450549
b =  29 369 0.810989010989011
b =  30 342 0.7516483516483516


In [32]:
b=27
accurate_rows = 0

for x, y in zip (X_binarised_train, y_train) :
        y_pred = (np.sum(x)>=b)
        accurate_rows += (y==y_pred)
print("b = ", b, accurate_rows, accurate_rows/X_binarised_train.shape[0])

b =  27 387 0.8505494505494505


In [33]:
b=27
accurate_rows = 0

for x, y in zip (X_binarised_train, y_train) :
        y_pred = (np.sum(x)>=b)
        accurate_rows += (y==y_pred)
print("b = ", b, accurate_rows, accurate_rows/X_binarised_train.shape[0])

b =  27 387 0.8505494505494505


In [34]:
b=25
accurate_rows = 0

for x, y in zip (X_binarised_train, y_train) :
        y_pred = (np.sum(x)>=b)
        accurate_rows += (y==y_pred)
print("b = ", b, accurate_rows, accurate_rows/X_binarised_train.shape[0])

b =  25 361 0.7934065934065934


In [35]:
for b in range (X_binarised_test.shape[1]+1):
    accurate_rows = 0
    for x, y in zip (X_binarised_test, y_train) :
        y_pred = (np.sum(x)>=b)
        accurate_rows += (y==y_pred)
    print("b = ", b, accurate_rows, accurate_rows/X_binarised_test.shape[0])

b =  0 77 0.6754385964912281
b =  1 77 0.6754385964912281
b =  2 77 0.6754385964912281
b =  3 77 0.6754385964912281
b =  4 77 0.6754385964912281
b =  5 76 0.6666666666666666
b =  6 76 0.6666666666666666
b =  7 77 0.6754385964912281
b =  8 76 0.6666666666666666
b =  9 76 0.6666666666666666
b =  10 76 0.6666666666666666
b =  11 76 0.6666666666666666
b =  12 76 0.6666666666666666
b =  13 75 0.6578947368421053
b =  14 73 0.6403508771929824
b =  15 71 0.6228070175438597
b =  16 70 0.6140350877192983
b =  17 69 0.6052631578947368
b =  18 69 0.6052631578947368
b =  19 64 0.5614035087719298
b =  20 63 0.5526315789473685
b =  21 64 0.5614035087719298
b =  22 63 0.5526315789473685
b =  23 62 0.543859649122807
b =  24 62 0.543859649122807
b =  25 61 0.5350877192982456
b =  26 60 0.5263157894736842
b =  27 54 0.47368421052631576
b =  28 55 0.4824561403508772
b =  29 51 0.4473684210526316
b =  30 48 0.42105263157894735


## Conclusion

When training set b = 25 , accurate rows= 361 and Got accuracy = 0.7934
Testing at accurate rows = 61 and Got accuracy = 0.53